modules/engine/src/shader-inputs.ts

```diff
@@ -0,0 +1,131 @@
+// luma.gl, MIT license
+import type {UniformValue, Texture, Sampler} from '@luma.gl/core';
+import {log} from '@luma.gl/core';
+// import type {ShaderUniformType, UniformValue, UniformFormat, UniformInfoDevice, Texture, Sampler} from '@luma.gl/core';
+import {_resolveModules, ShaderModuleInstance} from '@luma.gl/shadertools'
+
+
+/** Minimal ShaderModule subset, we don't need shader code etc */
+export type ShaderModuleInputs<
+  PropsT extends Record<string, unknown> = Record<string, unknown>,
+  UniformsT extends Record<string, UniformValue> = Record<string, UniformValue>,
+  BindingsT extends Record<string, Texture | Sampler> = Record<string, Texture | Sampler>
+> = {
+  defaultUniforms?: UniformsT;
+  getUniforms?: (settings: Partial<PropsT>, prevUniforms?: UniformsT) => UniformsT;
+
+  /** Not used. Used to access props type */
+  props?: PropsT;
+
+  bindings?: Record<
+    keyof BindingsT,
+    {
+      location: number;
+      type: 'texture' | 'sampler' | 'uniforms';
+    }
+  >;
+
+  uniformTypes?: any;
+};
+
+/**
+ * ShaderInputs holds a uniform and binding alues for one or more shader modules,
+ * - It can generate binary data for any uniform buffer
+ * - It can manage a uniform buffer for each block
+ * - It can update managed uniform buffers with a single call
+ * - It performs some book keeping on what has changed to minimize unnecessary writes to uniform buffers.
+ */
+export class ShaderInputs<
+  ShaderPropsT extends Partial<Record<string, Record<string, unknown>>> = Partial<
+    Record<string, Record<string, unknown>>
+  >
+> {
+  /** 
+   * The map of modules 
+   * @todo should should this include the resolved dependencies?
+   */
+  modules: Readonly<{[P in keyof ShaderPropsT]: ShaderModuleInputs<ShaderPropsT[P]>}>;
+
+  /** Stores the uniform values for each module */
+  moduleUniforms: Record<keyof ShaderPropsT, Record<string, UniformValue>>;
+  /** Stores the uniform bindings for each module  */
+  moduleBindings: Record<keyof ShaderPropsT, Record<string, Texture | Sampler>>;
+  /** Tracks if uniforms have changed */
+  moduleUniformsChanged: Record<keyof ShaderPropsT, false | string>;
+
+  /**
+   * Create a new UniformStore instance
+   * @param modules
+   */
+  constructor(modules: {[P in keyof ShaderPropsT]: ShaderModuleInputs<ShaderPropsT[P]>}) {
+    // TODO - get all dependencies from modules
```

Comment:
The question of where to fetch the dependencies is tricky and I'm also wondering if there is some clever way we can let Typescript resolve the dependency graph.

I did a [quick implementation](https://github.com/visgl/luma.gl/pull/1898/files#diff-906b8966317b75104205278a1cd0e0d8f2a694315c105f1685381c8a48d53070R62-R71) when trying out the `ShaderInputs` in the gltf example because I was not able to set the `lights` uniforms. 

Even with this change the types don't get infered and thus the `hello-gltf` app cannot know what it should be passing to `setProps`. This is because the model created for the gltf only has the `pbr` module listed, and the `lights` module is pulled in as a dependency.

Reply:
Yes. The application will just need to add the dependency to the ShaderInputs if it wants typed uniforms. I personally do not think that is too onerous, and would not try to lean in to heavily into trying to hide shader modules dependencies.

Basically, as soon as the app starts setting uniforms in a dependency module it by definitions knows about that dependency module (the props it sets must follow the docs of that module etc).